File: studentassistant/src/test/java/ru/erdenian/studentassistant/schedule/SemesterTest.java
package ru.erdenian.studentassistant.schedule;

import org.joda.time.LocalDate;
import org.junit.Before;
import org.junit.Test;

import static org.junit.Assert.assertEquals;

/**
 * Todo: описание класса
 *
 * @author Pavel Lavrikov
 *         *
 * @since Todo: текущая версия
 */
public class SemesterTest {
    private Semester semester;
    private String name;
    private LocalDate firstDay;
    private LocalDate lastDay;
    private long id;

    @Before
    public void initialize() {
        name = "5 семестр";
        firstDay = new LocalDate(2016, 9, 1);
        lastDay = new LocalDate(2016, 12, 10);
        id = System.nanoTime();
        semester = new Semester(name, firstDay, lastDay, id);
    }

    @Test
    public void getLength() {
        String tempName = "5 семестр";
        int fDay = 1;
        int lDay = 10;
        int difference = lDay - fDay + 1;
        LocalDate tempFirstDay = new LocalDate(2016, 10, fDay);
        LocalDate tempLastDay = new LocalDate(2016, 10, lDay);
        long tempId = System.nanoTime();
        Semester semesterTest = new Semester(tempName, tempFirstDay, tempLastDay, tempId);
        assertEquals("Возвращена неверная длина.", difference, semesterTest.getLength());

    }

    @Test
    public void compareToEquals() {
        String name = "5 семестр";
        LocalDate firstDay = new LocalDate(2016, 9, 1);
        LocalDate lastDay = new LocalDate(2016, 12, 10);
        long id = System.nanoTime();
        Semester semesterTest = new Semester(name, firstDay, lastDay, id);
        assertEquals("Сравнение с таким же объектом провалилось. Семестры не равны.", 0, semester.compareTo(semesterTest));
        //assertEquals("Семестры  не одинаковы.\nТест не пройден.",1,semester.compareTo(semesterTest));
        //assertTrue("Семестры одинаковы.\nТест пройден.",semester.compareTo(semesterTest)==0);
        //assertTrue("Семестры  не одинаковы.\nТест не пройден.",semester.compareTo(semesterTest)==1);
    }

    @Test
    public void compareToNotEquals() {
        String tempName = "6 семестр";
        LocalDate tempFirstDay = new LocalDate(2016, 10, 1);
        LocalDate tempLastDay = new LocalDate(2016, 12, 10);
        long tempId = System.nanoTime();
        Semester semesterTest = new Semester(tempName, tempFirstDay, tempLastDay, tempId);
        assertEquals("Сравнение с отличным объектом провалилось. Семестры равны.", -1, semester.compareTo(semesterTest));
        //assertEquals("Семестры  одинаковы.\nТест не пройден.",0,semester.compareTo(semesterTest));
        //assertTrue("Семестры  не одинаковы.\nТест пройден.",semester.compareTo(semesterTest)==1);
        //assertTrue("Семестры  одинаковы.\nТест не пройден.",semester.compareTo(semesterTest)==0);
    }

    @Test
    public void getName() {
        String tempName = semester.getName();
        assertEquals("Возвращено верное значение.\nТест пройден.", name, tempName);
    }

    @Test
    public void getFirstDay() {
        LocalDate tempFirstDay = semester.getFirstDay();
        assertEquals("Возвращено верное значение.\nТест пройден.", firstDay, tempFirstDay);
    }

    @Test
    public void getLastDay() {
        LocalDate tempLastDay = semester.getLastDay();
        assertEquals("Возвращено верное значение.\nТест пройден.", lastDay, tempLastDay);
    }

    @Test
    public void getId() {
        long tempId = semester.getId();
        assertEquals("Возвращено верное значение.\nТест пройден.", id, tempId);
    }

    @Test(expected = Exception.class)
    public void constructorEmptyName() {
        String tempName = "";
        LocalDate tempFirstDay = new LocalDate(2016, 10, 1);
        LocalDate tempLastDay = new LocalDate(2016, 12, 10);
        long tempId = System.nanoTime();
        Semester semesterTest = new Semester(tempName, tempFirstDay, tempLastDay, tempId);
    }

    @Test(expected = Exception.class)
    public void constructorIncorrectData() {
        String tempName = "";
        LocalDate tempFirstDay = new LocalDate(2016, 13, 1);
        LocalDate tempLastDay = new LocalDate(2016, 9, 10);
        long tempId = System.nanoTime();
        Semester semesterTest = new Semester(tempName, tempFirstDay, tempLastDay, tempId);
    }

}

File: features/repository/api/src/test/kotlin/ru/erdenian/studentassistant/repostory/api/entity/SemesterTest.kt
package ru.erdenian.studentassistant.repostory.api.entity

import java.time.LocalDate
import org.junit.Assert.assertEquals
import org.junit.Test
import ru.erdenian.studentassistant.repository.api.entity.Semester

class SemesterTest {

    @Test
    fun lengthTest() {
        assertEquals(
            1,
            Semester(
                name = "semester1",
                firstDay = LocalDate.of(2023, 2, 13),
                lastDay = LocalDate.of(2023, 2, 13),
                id = 1L,
            ).length,
        )
        assertEquals(
            2,
            Semester(
                name = "semester1",
                firstDay = LocalDate.of(2023, 2, 13),
                lastDay = LocalDate.of(2023, 2, 14),
                id = 1L,
            ).length,
        )
        assertEquals(
            365,
            Semester(
                name = "semester1",
                firstDay = LocalDate.of(2023, 2, 13),
                lastDay = LocalDate.of(2024, 2, 12),
                id = 1L,
            ).length,
        )
    }

    @Test
    fun dateRangeTest() {
        val start = LocalDate.of(2023, 2, 13)
        val end = LocalDate.of(2023, 5, 31)
        val semester = Semester(name = "semester1", firstDay = start, lastDay = end, id = 1L)

        assertEquals(start..end, semester.dateRange)
    }

    @Test
    fun getWeekNumberTest() {
        val semester1 = Semester(
            name = "semester1",
            firstDay = LocalDate.of(2023, 2, 13),
            lastDay = LocalDate.of(2024, 2, 12),
            id = 1L,
        )
        assertEquals(-1, semester1.getWeekNumber(LocalDate.of(2023, 2, 12)))
        assertEquals(0, semester1.getWeekNumber(LocalDate.of(2023, 2, 13)))
        assertEquals(0, semester1.getWeekNumber(LocalDate.of(2023, 2, 19)))
        assertEquals(1, semester1.getWeekNumber(LocalDate.of(2023, 2, 20)))
        assertEquals(1, semester1.getWeekNumber(LocalDate.of(2023, 2, 26)))

        val semester2 = Semester(
            name = "semester2",
            firstDay = LocalDate.of(2023, 2, 15),
            lastDay = LocalDate.of(2024, 2, 14),
            id = 1L,
        )
        assertEquals(-1, semester2.getWeekNumber(LocalDate.of(2023, 2, 12)))
        assertEquals(0, semester2.getWeekNumber(LocalDate.of(2023, 2, 13)))
        assertEquals(0, semester2.getWeekNumber(LocalDate.of(2023, 2, 19)))
        assertEquals(1, semester2.getWeekNumber(LocalDate.of(2023, 2, 20)))
        assertEquals(1, semester2.getWeekNumber(LocalDate.of(2023, 2, 26)))
    }

    @Test
    fun compareToEqualsTest() {
        val semester1 = Semester(
            name = "semester1",
            firstDay = LocalDate.of(2023, 2, 13),
            lastDay = LocalDate.of(2024, 2, 13),
            id = 1L,
        )
        val semester2 = Semester(
            name = "semester1",
            firstDay = LocalDate.of(2023, 2, 13),
            lastDay = LocalDate.of(2024, 2, 13),
            id = 1L,
        )
        assertEquals(0, semester1.compareTo(semester2))
    }

    @Test
    fun compareToLastDayTest() {
        val semester1 = Semester(
            name = "semester1",
            firstDay = LocalDate.of(2023, 2, 13),
            lastDay = LocalDate.of(2024, 2, 13),
            id = 1L,
        )
        val semester2 = Semester(
            name = "semester1",
            firstDay = LocalDate.of(2023, 2, 13),
            lastDay = LocalDate.of(2024, 2, 14),
            id = 1L,
        )
        assertEquals(-1, semester1.compareTo(semester2))
        assertEquals(1, semester2.compareTo(semester1))
    }

    @Test
    fun compareToFirstDayTest() {
        val semester1 = Semester(
            name = "semester1",
            firstDay = LocalDate.of(2023, 2, 13),
            lastDay = LocalDate.of(2024, 2, 13),
            id = 1L,
        )
        val semester2 = Semester(
            name = "semester1",
            firstDay = LocalDate.of(2023, 2, 14),
            lastDay = LocalDate.of(2024, 2, 13),
            id = 1L,
        )
        assertEquals(-1, semester1.compareTo(semester2))
        assertEquals(1, semester2.compareTo(semester1))
    }

    @Test
    fun compareToSubjectNameTest() {
        val semester1 = Semester(
            name = "semester1",
            firstDay = LocalDate.of(2023, 2, 13),
            lastDay = LocalDate.of(2024, 2, 13),
            id = 1L,
        )
        val semester2 = Semester(
            name = "semester2",
            firstDay = LocalDate.of(2023, 2, 13),
            lastDay = LocalDate.of(2024, 2, 13),
            id = 1L,
        )
        assertEquals(-1, semester1.compareTo(semester2))
        assertEquals(1, semester2.compareTo(semester1))
    }

    @Test
    fun compareToIdTest() {
        val semester1 = Semester(
            name = "semester1",
            firstDay = LocalDate.of(2023, 2, 13),
            lastDay = LocalDate.of(2024, 2, 13),
            id = 1L,
        )
        val semester2 = Semester(
            name = "semester1",
            firstDay = LocalDate.of(2023, 2, 13),
            lastDay = LocalDate.of(2024, 2, 13),
            id = 2L,
        )
        assertEquals(-1, semester1.compareTo(semester2))
        assertEquals(1, semester2.compareTo(semester1))
    }
}
